### ISSUE TITLE
Prometheus metrics

### ISSUE THREAD

**Issue by jpds** [bluejekyll/trust-dns#2032]
**Is your feature request related to a problem? Please describe.**

I currently use Unbound with https://github.com/letsencrypt/unbound_exporter/ (previously https://github.com/svartalf/unbound-telemetry) and am considering migrating my deployments to trust-dns.

However, I need the observability to keep an eye on things like query counts and latencies before users themselves start noticing them.

It'd be neat if trust-dns could integrate with https://crates.io/crates/prometheus-client to simply have native Prometheus metrics built into the server.

**Comment by djc** [comment]
This is for a server? Is there anything blocking you from implementing this outside the core trust-dns code?

It's not clear to me that a Prometheus integration should be delivered as part of this project.

**Comment by jpds** [comment]
Yes, this is for the server. There does not appear to be an internal mechanism to gather metrics within the trust-dns codebase and present those to an external exporter/service. Once that's in place, the majority of the work is done and all that's missing is an HTTP endpoint to present the metrics (+ an optional crate feature to enable them).

For C based projects (such as unbound), there's hesitation to natively integrate Prometheus as it requires a web endpoint for the metrics (which of course is a small security nightmare in C). Which is both why unbound has a number of different external exporter projects (in various stages of maintenance) and Go/Rust projects typically just natively integrate (for example: [Quickwit](https://quickwit.io/docs/reference/metrics/)).

**Comment by djc** [comment]
I think we currently don't have a lot of the infrastructure for this in place, but if you want to contribute, say, something to track statistics in a server, I'd be happy to provide guidance for that.

To me it feels like the server binary that trust-dns delivers is more of a prototype or tech preview and the project is mostly focused on libraries, so for now that's what I'd want to focus on.

**Comment by bluejekyll** [comment]
Just for reference, I think the way that we should consider integrating this is in two parts. We should probably look at embracing tokio-trace more deeply and then that has its own metrics collectors implemented for promethius.

So, use https://docs.rs/tracing/0.1.37/tracing/ in the libraries and then we could add a configuration to the server binary with a prometheus exporter https://docs.rs/metrics-exporter-prometheus/latest/metrics_exporter_prometheus/, I think those are all the correct things...

**Comment by scabala** [comment]
Hello,
I feel like this is going to be implemented because it was mentioned on #2725 (correct me if I'm wrong) but I would like to kindly, very gently ask: is there any time estimation when that might be done?

I'm currently comparing my options for self-hosted DNS server and lack of metrics is the _only_ weak point of HickoryDNS.

I do not know Rust (yet) so cannot help with implementation but I can offer testing such feature.

**Comment by djc** [comment]
Likely in the next 6 months-ish. Sooner if you can sponsor it.

**Comment by hargut** [comment]
Hi @divergentdave, @djc, @bluejekyll,

thank you as well for your support and collaboration! 👍
